onnx/checker.cc

```diff
@@ -216,25 +216,51 @@ void check_graph(const GraphProto& graph, int ir_version) {
   for (const auto& value_info : graph.output()) {
     check_value_info(value_info, ir_version);
   }
-  for (const auto& node : graph.node()) {
-    try {
-      check_node(node, ir_version);
-    } catch (ValidationError& ex) {
-      ex.AppendContext("Bad node spec: " + node.ShortDebugString());
-      throw ex;
-    }
-   }
 
-  std::unordered_set<std::string> input_names{};
+  std::unordered_set<std::string> output_names{};
   for (const auto& value_info : graph.input()) {
-    input_names.insert(value_info.name());
+    output_names.insert(value_info.name());
   }
   for (const auto& init : graph.initializer()) {
-    if (!input_names.count(init.name())) {
+    if (!output_names.count(init.name())) {
       fail_check(init.name() + " in initializer but not in graph input");
     }
     check_tensor(init, ir_version);
   }
+
+  for (const auto& node : graph.node()) {
+    // nodes must be in topologically sorted order
+    for (const auto& input : node.input()) {
+      // explicit optional input
+      if (input.empty()) {
+        continue;
+      }
+      if (!output_names.count(input)) {
+        fail_check(
+            "Nodes in a graph must be topologically sorted, however input '",
+            input,
+            "' of node: \n",
+            node.ShortDebugString(),
+            "\n is not output of any previous nodes.");
+      }
+    }
+    // check for SSA form
+    for (const auto& output : node.output()) {
+      if (output_names.count(output)) {
```

Comment:
output name could also be empty when it's optional and you may skip the check when it is.

Reply:
Good catch, fixed.